File: VSM Studio/drivers/RaspberryPi/timer.py
import wiringpi

class Timer:
   _period = -1
   _running = False
   _now = 0
   _time = 0
   _counter = 0
   _hook = None

   def initialize(self, period):
      self._period = period

   def setPeriod(self, period):
      if period > 0:
         self._period = period

   def start(self):
      if not self._running:
         self._time = self._now
         if (self._period > 0):
            self._running = True

   def stop(self):
      self._running = False

   def restart(self):
      self._counter = 0
      self.start()

   def resume(self):
      if (self._period > 0):
         self._running = True

   def read(self):
      return self._counter

   def attachInterrupt(self, hook, period=-1):
      self._hook = hook;
      self.setPeriod(period)

   # Event handler
   def _eventHandler(self):
      self._counter+=1
      # Restart
      self.start()
      if (self._hook):
         self._hook()

   def poll(self):
      self._now = wiringpi.millis()
      if self._running:
         passed = self._now - self._time
         if passed > self._period:
            self._time += self._period
            self._eventHandler()

File: VSM Studio/drivers/arduino/Controls/Timer.cpp
/********************************************************************/
/********************************************************************/
/*****                                                          *****/
/*****        L A B C E N T E R    E L E C T R O N I C S        *****/
/*****                                                          *****/
/*****                    Proteus IoT Builder                   *****/
/*****                                                          *****/
/*****                         Timer Control                    *****/
/*****                                                          *****/
/********************************************************************/
/********************************************************************/

#include "vfpserver.h"
#include "controls.h"

Timer::Timer (const char *id) : Control(id)
 { dir = 0;
   mode = 0;
   period = time = 0;
   running = triggered = 0;
 }


void Timer::begin (int8_t d, uint8_t m, uint8_t hour, uint8_t min, uint8_t sec, bool e)
 { dir = d;
   mode = m;
   period = time = sec+60L*min+3600L*hour;
   enabled = e;
   reset();
   nexttick = millis();
 } 

void Timer::poll ()
 { while ((long)(millis()-nexttick) >= 0)
    { if (running)
       { if (dir > 0 && time != period || dir < 0 && time != 0)
          { if (dir > 0 && ++time == period || dir < 0 && --time == 0)
               trigger();
            else
               sync();   
          }
       }
      nexttick += 1000;    
    } 
 }    

void Timer::start ()
 { if (dir > 0 && time == period || dir < 0 && time == 0)         
      reset(true);
   else   
    { running = true; 
      sync(); 
    }
 }

void Timer::reset (bool run)
 { running = run; 
   time = dir > 0 ? 0 : period;  
   sync();
 }

  
void Timer::setPeriod (uint8_t hour, uint8_t min, uint8_t sec)
 { period = sec+60L*min+3600L*hour;
   sync();
 }

void Timer::trigger()
 { if (mode == 0) // One-Shot
    { running = false;
      sync();
    }
   else
      restart();
   if (enabled)
    { triggered = true;      
      if (eventHandler != NULL)    
         eventHandler();         
      VFP.call(id, F("trigger")).end();         
    }      
 }
 
void Timer::sync ()
 { VFP.call(id, F("sync")).arg(hour(time)).arg(minute(time)).arg(second(time)).arg(enabled).end();
 }